### PR TITLE
Publish solutions for lab 8

### PR DESCRIPTION
**Pull request by sketch873**

### CODE REVIEW

File: labs/lab08/sol/lin/6-once/once.c [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+
+	for (i = 0; i < NUM_THREADS; i++) {
+        rc = pthread_create(&threads[i], NULL, thread_func, (void *) i);
+	    DIE(rc != 0, "pthread_create");
+    }
+
+	for (i = 0; i < NUM_THREADS; i++) {
+	    rc = pthread_join(threads[i], NULL);
+	    DIE(rc != 0, "pthread_join");
+    }
```
Comment (teodutu) [inline labs/lab08/sol/lin/6-once/once.c]:
```suggestion
		rc = pthread_join(threads[i], NULL);
		DIE(rc != 0, "pthread_join");
	}
```

File: labs/lab08/sol/lin/6-once/once.c [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+	pthread_t threads[NUM_THREADS];
+	long i;
+	int rc;
+
+	/* Create NUM_THREADS threads, all will call one_time_init() / one time deinit */
+
+	for (i = 0; i < NUM_THREADS; i++) {
+        rc = pthread_create(&threads[i], NULL, thread_func, (void *) i);
+	    DIE(rc != 0, "pthread_create");
+    }
```
Comment (teodutu) [inline labs/lab08/sol/lin/6-once/once.c]:
```suggestion
		rc = pthread_create(&threads[i], NULL, thread_func, (void *) I);
		DIE(rc != 0, "pthread_create");
	}
```

File: labs/lab08/sol/lin/6-once/once.c [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+   */
+
+	rc = pthread_mutex_lock(&(once_control->mutex));
+	DIE(rc == -1, "pthread_mutex_lock");
+
+	once_control->refcount--;
+
+	if (!once_control->refcount) {
+		(*f)();
+	}
```
Comment (teodutu) [inline labs/lab08/sol/lin/6-once/once.c]:
```suggestion
	if (!once_control->refcount)
		(*f)();
```

File: labs/lab08/sol/lin/6-once/once.c [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+	return 0;
+}
+
+static int one_time_deinit(struct once_struct *once_control, void (*f)(void))
+{
+	int rc;
+
+	/* TODO - Each time a thread gets here, decrement once_control->refcount.
+   * Call the f() function only if this is the last thread that got here.
+   */
```
Comment (teodutu) [inline labs/lab08/sol/lin/6-once/once.c]:
```suggestion
	* Call the f() function only if this is the last thread that got here.
	*/
```

Do this to `lab08/skel/lin/6-once/once.c` as well.

File: labs/lab08/sol/lin/6-once/once.c [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+    }
+}
+
+static int one_time_init(struct once_struct *once_control, void (*f)(void))
+{
+	int rc;
+
+	/* TODO - Each time a thread gets here, increment once_control->refcount.
+   * Call the f() function only if this is the first thread that got here.
+   */
```
Comment (teodutu) [inline labs/lab08/sol/lin/6-once/once.c]:
```suggestion
	* Call the f() function only if this is the first thread that got here.
	*/
```

Do this to `lab08/skel/lin/6-once/once.c` as well.

File: labs/lab08/sol/lin/6-once/once.c [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+
+	rc = pthread_mutex_lock(&(once_control->mutex));
+	DIE(rc == -1, "pthread_mutex_lock");
+
+	if (!once_control->refcount) {
+		(*f)();
+	}
```
Comment (teodutu) [inline labs/lab08/sol/lin/6-once/once.c]:
```suggestion
	if (!once_control->refcount)
		(*f)();
```

File: labs/lab08/sol/lin/6-once/once.c [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+        printf("%c", START_CHAR + arg);
+
+        /* What happens if you comment out the fflush ? */
+
+        fflush(stdout);
+
+        /* What happens if you decrease the PERIOD ? */
+
+        usleep(PERIOD);
+    }
```
Comment (teodutu) [inline labs/lab08/sol/lin/6-once/once.c]:
```suggestion
	for (int i = 0; i < NUM_CHARS; i++) {
		printf("%c", START_CHAR + arg);

		/* What happens if you comment out the fflush? */

		fflush(stdout);

		/* What happens if you decrease the PERIOD? */

		usleep(PERIOD);
	}
```

Apply this to `lab08/skel/lin/6-once/once.c` as well.

File: labs/lab08/sol/lin/6-once/once.c [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+	 */
+
+	time_t ts = time(NULL);
+
+	printf("Threads execution started at %s", asctime(gmtime(&ts)));
+}
+
+static void deinit_func(void)
+{
+
```
Comment (teodutu) [inline labs/lab08/sol/lin/6-once/once.c]:
```suggestion
```

File: labs/lab08/sol/lin/3-safety/malloc.c [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+#define NUM_COLUMNS 1000
+
+#define PRINT_HDR_SZ    32
+#define PRINT_BUF_SZ    256
+#define MAIN_PRINT      (-1)
+
+int global_storage;
+
+static void increase_numbers(int *a, int *b, int*c) {
+	(*a)++;
```
Comment (teodutu) [inline labs/lab08/sol/lin/3-safety/malloc.c]:
```suggestion
static void increase_numbers(int* a, int* b, int* c)
{
	(*a)++;
```

Do the same for `lab08/skel/lin/3-safety/malloc.c`.

File: labs/lab08/sol/lin/3-safety/malloc.c [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+	(*b)++;
+	(*c)++;
+}
+
+static void print_stats(int origin,
+						int *global_storage,
+						int *function_global_storage,
+						int *function_specific_storage)
```
Comment (teodutu) [inline labs/lab08/sol/lin/3-safety/malloc.c]:
```suggestion
			int *global_storage,
			int *function_global_storage,
			int *function_specific_storage)
```

It just looked weird like that.

File: labs/lab08/sol/lin/3-safety/critical.c [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+		 *
+		 * HINT 2: Use the most restrictive memory order for the
+		 * second parameter of __atomic_test_and_set. See the
+		 * https://gcc.gnu.org/onlinedocs/gcc/_005f_005fatomic-Builtins.html
+		 * link and search for "__ATOMIC_SEQ_CST"
+		 */
+
+		if(__atomic_test_and_set(&lock, __ATOMIC_SEQ_CST) == false) {
+			return;
+		}
```
Comment (teodutu) [inline labs/lab08/sol/lin/3-safety/critical.c]:
```suggestion
		if (__atomic_test_and_set(&lock, __ATOMIC_SEQ_CST) == false)
			return;
```

File: labs/lab08/sol/lin/3-safety/Makefile [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+
+critical_spin.o: critical.c
+	$(CC) -c $(CFLAGS) $(CPPFLAGS) --std=c99 -DUSE_SPINLOCK -o $@ $<
+
+critical.o: critical.c
+	$(CC) -c $(CFLAGS) $(CPPFLAGS) -o $@ $<
+
+clean:
+	-rm -f *.o *~ spin mutex
+
```
Comment (teodutu) [inline labs/lab08/sol/lin/3-safety/Makefile]:
```suggestion
```

File: labs/lab08/sol/lin/3-safety/critical.c [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+static bool lock;
+#else
+static pthread_mutex_t lock = PTHREAD_MUTEX_INITIALIZER;
+#endif
+
+void acquire_lock(void)
+{
+#ifdef USE_SPINLOCK
+	/* Busy waiting */
+	while(1) {
```
Comment (teodutu) [inline labs/lab08/sol/lin/3-safety/critical.c]:
```suggestion
	while (1) {
```

The same for `lab08/skel/lin/3-safety/critical.c`.

File: labs/lab08/sol/lin/2-th_vs_proc/threads.c [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+
+int main(void)
+{
+	pthread_t thread[NUM_TASKS];
+	int th_id[NUM_TASKS];
+	int i, rc;
+
+	for (i = 0; i < NUM_TASKS; i++) {
+		th_id[i] = i;
+	/* Create a thread and assign do_task as its starting function */
```
Comment (teodutu) [inline labs/lab08/sol/lin/2-th_vs_proc/threads.c]:
```suggestion
		/* Create a thread and assign do_task as its starting function */
```

**Review by teodutu — CHANGES_REQUESTED** [review]
Apart from some coding style issues, everything looks in order.
Apart from what I wrote as inline code suggestions, please replace `lin` with `Linux` in the block comments of all Linux files.

**Review by teodutu — DISMISSED** [review]
I left 3 unresolved comments from before. The changes to the `sol/` folder are fine. Please apply them to the `skel` folder as well.

File: labs/lab08/sol/lin/7-prodcons/prodcons.c [hunk 15]
```diff
@@ -0,0 +1,188 @@
...
+			sleep(delay);
+	}
+
+	return NULL;
+}
+
+void run_threads(int producer_delay, int consumer_delay)
+{
+	int rc;
+	pthread_t producer_th, consumer_th;
```
Comment (andreia-oca) [inline labs/lab08/sol/lin/7-prodcons/prodcons.c]:
```suggestion
	pthread_t producer_th;
	pthread_t consumer_th;
```

File: labs/lab08/sol/lin/7-prodcons/prodcons.c [hunk 16]
```diff
@@ -0,0 +1,188 @@
...
+		else
+			sleep(delay);
+	}
+
+	return NULL;
+}
+
+void *consumer_fn(void *arg)
+{
+	int item_consumed, i, rc;
```
Comment (andreia-oca) [inline labs/lab08/sol/lin/7-prodcons/prodcons.c]:
```suggestion
	int item_consumed;
	int i;
	int rc;
```

File: labs/lab08/sol/lin/2-th_vs_proc/processes.c [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+void do_bad_task(void)
+{
+	printf("Process %d doing BAD task\n", getpid());
+
+	exit(-1);
+}
+
+int main(void)
+{
+	int i, rc;
```
Comment (andreia-oca) [inline labs/lab08/sol/lin/2-th_vs_proc/processes.c]: One declaration per line here also.

File: labs/lab08/sol/lin/2-th_vs_proc/threads.c [hunk 18]
```diff
@@ -0,0 +1,65 @@
...
+	printf("Thread %d doing BAD task\n", thread_no);
+
+	exit(-1);
+}
+
+int main(void)
+{
+	pthread_t thread[NUM_TASKS];
+	int th_id[NUM_TASKS];
+	int i, rc;
```
Comment (andreia-oca) [inline labs/lab08/sol/lin/2-th_vs_proc/threads.c]:
```suggestion
	int i;
	int rc;
```

File: labs/lab08/sol/lin/7-prodcons/prodcons.c [hunk 19]
```diff
@@ -0,0 +1,188 @@
...
+buffer_t common_area;
+
+pthread_cond_t buffer_not_full  = PTHREAD_COND_INITIALIZER;
+pthread_cond_t buffer_not_empty = PTHREAD_COND_INITIALIZER;
+
+pthread_mutex_t mutex = PTHREAD_MUTEX_INITIALIZER;
+
+void *producer_fn(void *arg)
+{
+	int item_to_insert, i, rc;
```
Comment (andreia-oca) [inline labs/lab08/sol/lin/7-prodcons/prodcons.c]:
```suggestion
	int item_to_insert;
	int i;
	int rc;
```

**Review by andreia-oca — DISMISSED** [review]
Looks good to me. 
I have just one general comment. The recommended coding style is one declaration per line, no commas for multiple declarations). I added suggestions in the source code where I spotted this problem.

**Review by teodutu — APPROVED** [review]

**Review by andreia-oca — APPROVED** [review]

**Pull request merged**